.github/workflows/tests.yml

```diff
@@ -1,27 +1,106 @@
 ---
-name: Rust
+name: Tests
 on:
   push:
     branches: [main]
   pull_request:
     branches: [main]
 
-env:
-  CARGO_TERM_COLOR: always
-  ACTIONS_LINTS_TOOLCHAIN: 1.50.0
-
 jobs:
-  linting:
-    name: "Lints, pinned toolchain"
+  tests:
+    # Distro configuration matrix
+    #
+    # Each build is run in a Docker container specific to the distro.
+    # When adding a new distro, handle the dependency installation in
+    # `ci/gh-install.sh`. The matrix configuration options are:
+    #
+    # name: A friendly name to use for the job.
+    #
+    # image: The Docker image to use.
+    #
+    # pre-checkout-setup: Commands to run before the git repo checkout.
+    #   If git is not in the Docker image, it must be installed here.
+    #   Otherwise, the checkout action uses the GitHub REST API, which
+    #   doesn't result in an actual git repo. A real git repo is
+    #   required to checkout the submodules.
+    #
+    # extra-packages: Packages to install in addition to those in
+    #   `ci/gh-install.sh`. This can be used to support features from
+    #   additional `configure` options.
+    #
+    # configure-options: Options to pass to `configure`.
+    strategy:
+      matrix:
+        include:
+          - name: Debian Buster (10) with sign-ed25519
+            image: debian:buster-slim
+            pre-checkout-setup: |
+              apt-get update
+              apt-get install -y git
+            extra-packages: >-
+              libsodium-dev
+            configure-options: >-
+              --with-ed25519-libsodium
+
+          - name: Debian Buster (10) with curl, sign-ed25519 and no gpgme
+            image: debian:buster-slim
+            pre-checkout-setup: |
+              apt-get update
+              apt-get install -y git
+            extra-packages: >-
+              libsodium-dev
+            configure-options: >-
+              --with-curl
+              --with-ed25519-libsodium
+              --without-gpgme
+
+          # A 32 bit build to act as a proxy for frequently deployed 32
+          # bit armv7
+          - name: Debian Buster (10) 32 bit
+            image: i386/debian:buster-slim
+            # This is pretty nasty. The checkout action uses an x86_64
+            # node binary in the container, so we need to provide an
+            # x86_64 ld.so and libstdc++.
+            pre-checkout-setup: |
+              dpkg --add-architecture amd64
+              apt-get update
+              apt-get install -y git libc6:amd64 libstdc++6:amd64
+
+          - name: Ubuntu Focal (20.04)
+            image: ubuntu:focal
+            pre-checkout-setup: |
+              apt-get update
+              apt-get install -y git
+
+          - name: Ubuntu Groovy (20.10)
```

Comment:
groovy is a short-term-support release and has already been superseded by hirsute (21.04), so it might make more sense to go for hirsute - or Debian bullseye, which is roughly contemporary.

Reply:
Yeah, this was just a straight copy of the travis configuration. I agree we could target different releases here.

I actually think it would be better to use release phase names. I.e., stable and testing (and maybe even unstable) for Debian. The docker repo has those tags. Similarly, Ubuntu has latest pointing to the current LTS (focal), rolling pointing to the current short term release (hirsute) and devel pointing to the unreleased codename (impish).

We could certainly do all of that here, but I kind of wanted to just convert the setup as is.

Reply:
> I actually think it would be better to use release phase names. I.e., stable and testing (and maybe even unstable) for Debian

There isn't much point in doing both testing and unstable; they're meant to always be close, except perhaps during a freeze or a big transition. I'd recommend testing over unstable if you just want "something newish", because unstable sometimes has transient brokenness (uninstallability etc.) which the migration process from unstable to testing is designed to filter out.